drivers: entropy: stm32: make `get_entropy_isr` re-entrant

mathieuchopstm opened the pull request:
The current `get_entropy_isr` implementation is not re-entrant; in fact, it will always deadlock when a re-entrant call is performed, because the logic to determine whether a call should disable the RNG or not after completion is inverted (the RNG is "acquired" when its interrupt is **DISABLED**, not when it is **ENABLED**) ==> re-entrant calls will always disable the RNG before return. Depending on where the original/pre-empted call was when it was interrupted, this can lead to the function polling the (now disabled) RNG for entropy that will never arrive. Inverting the logic seems like a fix, but is actually incomplete as there are situations where this information is not sufficient alone to determine the proper action to take.

Protect the RNG using a reference count and modify `get_entropy_isr()` to take advantage of this and truly support re-entrant calls.

etienne-lms reviewed (commented):
The implementation seems quite complex. I fear this may hide issues.

mathieuchopstm committed:
drivers: entropy: stm32: reorganize includes

The includes section was a mess - reorganize it.

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>

mathieuchopstm committed:
drivers: entropy: stm32: make get_entropy_isr() re-entrant

The current `get_entropy_isr()` implementation is not re-entrant; in fact,
re-entrant calls have a very high chance of deadlocking today because the
logic that determines whether we should disable the RNG after generating
entropy is inverted: the RNG is considered as "acquired" for ISR polling
when its interrupt is ENABLED, but get_entropy_isr() DISABLES the IRQ...

Inverting the logic seems like it can fix the problem, but the logic would
still be flawed: IRQ enablement is insufficient alone to handle all cases.

Protect the RNG enablement using a usage count (protected by spinlock) to
ensure that (1) the RNG is enabled atomically and (2) we never disable the
RNG at a wrong time, which allows get_entropy_isr() to work re-entrantly
with a few modifications.

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>
